File: cci/chapter_1/1.1.py

""" Is unique: Implement an algorithm to determine if a string has
    all unique characters. What if you cannot use additional data structures."""

# input - string of (n) length
# output - boolean

# assumptions - using ASCII, which is a 128 character set

# Naive approach using a set
# Time: O(n^2)
def all_unique_set(s):
    if len(s) > 128:
        return False

    seen = set()
    for char in s:
        if char in seen:
            return False
        chars.add(char)

    return True


# Using an array of boolean values mapped to chars
# Time: O(n)
def all_unique_array(s):
    if len(s) > 128:
        return False

    char_list = [False] * 128
    for char in s:
        idx = ord(char)
        if char_list[idx]:
            return False

        chars_list[idx] = True

    return True


# test cases
print(all_unique_set('abcdefg'))
print(all_unique_set('abcdefgg'))

print(all_unique_array('abcdefg'))
print(all_unique_array('abcdefgg'))

File: cci/js/chapter_1/questions/1.1.js

// Is unique: Implement an algorithm to determine if a string has
// all unique characters. What if you cannot use additional data structures.

// input: string of (n) length
// output: boolean

// assumptions - using ASCII, which is a 128 character set

// Using an array of boolean values mapped to chars
// Time: O(n)

const isUnique = function(str) {
  if (str.length > 128) {
    return false;
  }

  const chars = [];
  for (let i = 0; i < 128; i++) {
    chars[i] = false;
  }

  for (let i = 0; i < str.length; i++) {
    const idx = str.charCodeAt(i);

    if (chars[idx]) {
      return false;
    }
    chars[idx] = true;
  }

  return true;
}

// test cases
console.log(isUnique('abc')); // true
console.log(isUnique('abcc')); // false

